OCast/src/main/java/org/ocast/dial/DialApplication.kt

```diff
@@ -0,0 +1,200 @@
+         * </service>
+         *
+         * @param xml The XML string to decode.
+         * @return The decoded application.
+         * @throws Exception If an error occurs while decoding the DIAL application.
+         */
+        @Throws(Exception::class)
+        fun decode(xml: String): DialApplication<OCastAdditionalData> {
+            val rootXMLElement = XMLParser.parse(xml)
+            val serviceXMLElement = rootXMLElement[XML_SERVICE_ELEMENT_NAME]
+            val name = serviceXMLElement[XML_NAME_ELEMENT_NAME].value
+            val isStopAllowed = serviceXMLElement[XML_OPTIONS_ELEMENT_NAME].attributes[XML_ALLOW_STOP_ATTRIBUTE_NAME] == XML_TRUE_TEXT_VALUE
+            val state = decodeState(serviceXMLElement[XML_STATE_ELEMENT_NAME].value)
+            val instancePath = serviceXMLElement
+                .getOrNull(XML_LINK_ELEMENT_NAME)
+                ?.attributes
+                ?.get(XML_HREF_ATTRIBUTE_NAME)
+            val additionalDataXMLElement = serviceXMLElement[XML_ADDITIONAL_DATA_ELEMENT_NAME]
+            val webSocketURL = additionalDataXMLElement
+                .getOrNull(XML_OCAST_APP_2_APP_URL_ELEMENT_NAME)
+                ?.run {
+                    runCatching { URI(value) }.getOrNull()
+                }
+            val version = additionalDataXMLElement
+                .getOrNull(XML_OCAST_VERSION_ELEMENT_NAME)
+                ?.value
+
+            return DialApplication(name, isStopAllowed, state, instancePath, OCastAdditionalData(webSocketURL, version))
+        }
+
+        /**
+         * Decodes a raw state to a [State].
+         *
+         * @param rawState The raw state to decode.
+         * @return The decoded state.
+         * @throws DialError If the state is unknown.
+         * @throws MalformedURLException If state is [State.Installable] but URL is malformed.
+         */
+        @Throws(DialError::class, MalformedURLException::class)
+        private fun decodeState(rawState: String): State {
+            return when (rawState) {
+                "running" -> State.Running
+                "stopped" -> State.Stopped
+                "hidden" -> State.Hidden
+                else -> {
+                    val installationUrlString = try {
+                        rawState.split("installable=").elementAt(1)
+                    } catch (exception: Exception) {
+                        throw DialError("Unknown DIAL application state")
+                    }
+                    State.Installable(URL(installationUrlString))
+                }
+            }
+        }
+    }
+}
+
+/**
+ * A base class that represents additional data in a DIAL application information response.
+ */
+internal open class DialAdditionalData
+
+/**
+ * This class represents the additional data related to OCast when receiving DIAL application information response.
+ *
+ * @param webSocketURL The OCast web socket URL.
+ * @param version The OCast version.
+ */
+internal data class OCastAdditionalData(
```

Comment:
move OCastAdditionalData to org.ocast.dial.models package ?

Reply:
Fixed.
Moved `DialApplication` too.